### PR TITLE
Refactor codegen for long-running operations.

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
LROs will now return a future allowing the caller to get status on the
operation and control polling.  The Go routines have also been removed.

### CODE REVIEW

**Comment by olydis** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

**Comment by jhendrixMSFT** [comment]
Example codegen https://github.com/jhendrixMSFT/azure-sdk-for-go/commit/f5f62f3b51056bd640076c14a0ba112e182f3bde

File: src/Model/FutureTypeGo.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    /// <summary>
+    /// Represents a future, which is the return type for long-running operations.
+    /// </summary>
+    class FutureTypeGo : CompositeTypeGo
+    {
+        /// <summary>
+        /// Creates a new future type for the specified method.
+        /// </summary>
+        /// <param name="method">The method that will return a future.</param>
+        public FutureTypeGo(MethodGo method) : base($"{method.Name}ResponseFuture")
```
Comment (jhendrixMSFT) [inline src/Model/FutureTypeGo.cs]: Unfortunately this name can collide with other futures, will fix.

File: src/Templates/MethodTemplate.cshtml [hunk 2]
```diff
@@ -161,33 +131,52 @@ func (client @(Model.Owner)) @(Model.PreparerMethodName)(@(Model.MethodParameter
...
 
 @EmptyLine
 // @(Model.SenderMethodName) sends the @(Model.Name) request. The method will close the
 // http.Response Body if it receives an error.
-func (client @(Model.Owner)) @(Model.SenderMethodName)(req *http.Request) (*http.Response, error) {
+func (client @(Model.Owner)) @(Model.SenderMethodName)(req *http.Request) (@senderRetType, error) {
+@if (Model.IsLongRunningOperation())
+{
+<text>
+    sender := autorest.DecorateSender(client, azure.DoRetryWithRegistration(client.Client))
```
Comment (mcardosos) [inline src/Templates/MethodTemplate.cshtml]: `DoRetryWithRegistration` should not be used in all cases

Comment (jhendrixMSFT) [reply src/Templates/MethodTemplate.cshtml]: Good point, this should be predicated on Model.RegisterRP.

File: src/Model/FutureTypeGo.cs (hunk 1 above)
Comment (jhendrixMSFT) [reply src/Model/FutureTypeGo.cs]: Fixed.

File: src/Templates/MethodTemplate.cshtml (hunk 2 above)
Comment (jhendrixMSFT) [reply src/Templates/MethodTemplate.cshtml]: Fixed.

File: src/Templates/MethodTemplate.cshtml [hunk 3]
```diff
@@ -148,10 +148,14 @@ func (client @(Model.Owner)) @(Model.PreparerMethodName)(@(Model.MethodParameter
 func (client @(Model.Owner)) @(Model.SenderMethodName)(req *http.Request) (@senderRetType, error) {
 @if (Model.IsLongRunningOperation())
 {
+    var sender = "client";
+    if (Model.RegisterRP)
+    {
+        sender = "autorest.DecorateSender(client, azure.DoRetryWithRegistration(client.Client))";
+    }
```
Comment (mcardosos) [inline src/Templates/MethodTemplate.cshtml]: `DoRetryForStatusCodes` is missing when not registering RPs

Comment (jhendrixMSFT) [reply src/Templates/MethodTemplate.cshtml]: Fixed.

**Pull request merged**